### PR TITLE
ci: add automated hole-punch tests

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Related: https://github.com/libp2p/test-plans/pull/304.

## Attributions

Co-authored-by: diegomrsantos <diegomrsantos@gmail.com>

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: .github/workflows/interop-test.yml
Comment (thomaseizinger) [inline .github/workflows/interop-test.yml]: Should point to upstream once merged.

**Comment by thomaseizinger** [comment]
Once this is merged and released, we need to update the hash that is referenced in the `test-plans` Makefile to point to a new hash.

**Comment by diegomrsantos** [comment]
@thomaseizinger is this still relevant?

**Review by mxinden — APPROVED** [review]
Work looks good to me. Feel free to merge once CI is green :tada:

**Comment by diegomrsantos** [comment]
I see the commit hash in the `test-plans` is coming from here and that's the code downloaded into the `impl` dir. Should we also add support to the arm64 platform before merging?

**Comment by thomaseizinger** [comment]
> before merging?

First, we need to figure out why the tests are failing :D
I am not sure if our current image caching strategy supports multi-arch containers at all? I didn't write any of https://github.com/libp2p/test-plans/blob/1e37b93e939c75241dba726a5cb3be1cbb7a88a9/hole-punch-interop/helpers/cache.ts#L76-L125 but we'd have to investigate whether `docker image save` actually does the right thing for multi-arch images.

I'd like to not block this PR on it if possible. The state in `test-plans` is currently "dirty" because it points to a commit hash in here and I didn't plan on leaving that for very long. We should point to a (released) commit on `master` instead. Support for other architectures can be added later.

**Comment by thomaseizinger** [comment]
The failure is definitely not a flake, for some reason the listener (which is not this branch) on QUIC loses the connection to the relay. See https://github.com/libp2p/rust-libp2p/actions/runs/6565723299/job/17845062912?pr=4549#step:5:612.

**Comment by diegomrsantos** [comment]
@thomaseizinger l I've also added the new changes from https://github.com/libp2p/test-plans/pull/311 in  https://github.com/libp2p/rust-libp2p/pull/4687. Feel free to cherry-pick the commits here or handle the latter separately. I've tested everything locally and it seems to work, I don't have permission to trigger the CI.

> I am not sure if our current image caching strategy supports multi-arch containers at all?

I added support for building according to the current machine architecture. It should impact local buildings when someone is using arm64, but shouldn't change anything on your CI if it doesn't use arm64. Does it make sense?

File: transports/quic/src/hole_punching.rs
Comment (thomaseizinger) [inline transports/quic/src/hole_punching.rs:28]: @mxinden Sleeping here caused the tests to be quite flaky and probably impacted our real-world hole-punch success rate a lot as we effectively mitigated the `SYNC` part of DCUtR ..

File: hole-punching-tests/src/main.rs
Comment (thomaseizinger) [inline hole-punching-tests/src/main.rs:164]: This is a bit of a hack. For some reason, the underlying connection to the _relay_ can sometimes fail during the reservation. I'll create a follow-up ticket to investigate this later but for now, we can just reconnect here.

**Comment by thomaseizinger** [comment]
@mxinden This is ready for a re-review. I think I've ironed out the flakiness.

File: transports/quic/src/hole_punching.rs
Comment (mxinden) [reply transports/quic/src/hole_punching.rs:28]: Changing the order here is fine by me.

In general, I would expect a UDP hole punch to not need the time synchronization of DCUtR. Given that UDP is connectionless and thus NATs can not track the state of a QUIC connection, I would expect timing to matter less.

All that said, I don't see a reason why to sleep first and then send the message. Sending first makes a lot more sense, even if just for faster hole punching.

Good catch! Thanks for debugging!

**Review by mxinden — APPROVED** [review]

**Pull request merged**
